ImageGallery/src/org/sleuthkit/autopsy/imagegallery/datamodel/DrawableDB.java

```diff
@@ -766,6 +779,112 @@ public void insertFile(DrawableFile f, DrawableTransaction tr, CaseDbTransaction
     public void updateFile(DrawableFile f, DrawableTransaction tr, CaseDbTransaction caseDbTransaction) {
         insertOrUpdateFile(f, tr, updateFileStmt, caseDbTransaction);
     }
+    
+    
+    /**
+     * Populate caches based on current state of Case DB
+     */
+    synchronized public void buildFileMetaDataCache() {
+        cacheBuildCount++;
+        if (areCachesLoaded == true)
+            return;
+
+        try {
+            // get tags
+            try (SleuthkitCase.CaseDbQuery dbQuery = tskCase.executeQuery("SELECT obj_id FROM content_tags")) {
+                ResultSet rs = dbQuery.getResultSet();
+                while (rs.next()) {
+                    long id = rs.getLong("obj_id");
+                    hasTagCache.add(id);
+                }
+            } catch (SQLException ex) {
+                logger.log(Level.SEVERE, "Error getting tags from DB", ex); //NON-NLS
+            }
+        } catch (TskCoreException ex) {
+            logger.log(Level.SEVERE, "Error executing query to get tags", ex); //NON-NLS
+        }
+        
+        try {
+            // hash sets
+            try (SleuthkitCase.CaseDbQuery dbQuery = tskCase.executeQuery("SELECT obj_id FROM blackboard_artifacts WHERE artifact_type_id = " + BlackboardArtifact.ARTIFACT_TYPE.TSK_HASHSET_HIT.getTypeID())) {
+                ResultSet rs = dbQuery.getResultSet();
+                while (rs.next()) {
+                    long id = rs.getLong("obj_id");
+                    hasHashCache.add(id);
+                }
+                
+            } catch (SQLException ex) {
+                logger.log(Level.SEVERE, "Error getting hashsets from DB", ex); //NON-NLS
+            }
+        } catch (TskCoreException ex) {
+            logger.log(Level.SEVERE, "Error executing query to get hashsets", ex); //NON-NLS
+        }
+        
+        try {
+            // EXIF
+            try (SleuthkitCase.CaseDbQuery dbQuery = tskCase.executeQuery("SELECT obj_id FROM blackboard_artifacts WHERE artifact_type_id = " + BlackboardArtifact.ARTIFACT_TYPE.TSK_METADATA_EXIF.getTypeID())) {
+                ResultSet rs = dbQuery.getResultSet();
+                while (rs.next()) {
+                    long id = rs.getLong("obj_id");
+                    hasExifCache.add(id);
+                }
+                
+            } catch (SQLException ex) {
+                logger.log(Level.SEVERE, "Error getting EXIF from DB", ex); //NON-NLS
+            }
+        } catch (TskCoreException ex) {
+            logger.log(Level.SEVERE, "Error executing query to get EXIF", ex); //NON-NLS
+        }
+        
+        areCachesLoaded = true;
+    }
+    
+    /**
+     * Add a file to cache of files that have EXIF data
+     * @param objectID ObjId of file with EXIF
+     */
+    public void addExifCache(long objectID) {
```

Comment:
shouldn't the addXXXCache methods be synchronized too?

Reply:
Added syncronization in latest commit.